### PR TITLE
select built-in backend as you wish, can disable EGL, GLES, GBM or libinput

### PR DESCRIPTION
**Pull request by hizukiayaka**
Some SoCs don't have a GPU
Nvidia doesn't support gbm
Some products don't need input device, it would only display something

More option and flexible can be achieve as long as the more allocators and software renderer are merged.

### CODE REVIEW

File: include/wlr/config.h.in [hunk 1]
```diff
@@ -15,4 +15,9 @@
 
 #mesondefine WLR_HAS_XDG_FOREIGN
 
+#mesondefine WLR_HAS_GBM_SUPPORT
+#mesondefine WLR_HAS_GLES2_RENDERER
+
+#mesondefine WLR_HAS_DRM_BACKEND
+#mesondefine WLR_HAS_LIBINPUT
```
Comment (emersion) [inline include/wlr/config.h.in:22]: This should be named `WLR_HAS_LIBINPUT_BACKEND` for consistency with the rest of the backends.

File: render/meson.build [hunk 2]
```diff
@@ -1,13 +1,35 @@
...
-	'gbm_allocator.c',
 	'shm_format.c',
 	'swapchain.c',
 	'wlr_renderer.c',
 	'wlr_texture.c',
 )
+wlr_deps += drm_headers
+
+# EGLStream is not supported yet, without GBM the EGL would lost most of its
+# functions
```
Comment (emersion) [inline render/meson.build:17]: We will never support EGLStreams.

File: meson_options.txt [hunk 3]
```diff
@@ -4,7 +4,8 @@ option('libseat', type: 'feature', value: 'auto', description: 'Enable support f
 option('xcb-errors', type: 'feature', value: 'auto', description: 'Use xcb-errors util library')
 option('xcb-icccm', type: 'feature', value: 'auto', description: 'Use xcb-icccm util library')
 option('xwayland', type: 'feature', value: 'auto', yield: true, description: 'Enable support for X11 applications')
-option('x11-backend', type: 'feature', value: 'auto', description: 'Enable X11 backend')
 option('examples', type: 'boolean', value: true, description: 'Build example applications')
 option('icon_directory', description: 'Location used to look for cursors (default: ${datadir}/icons)', type: 'string', value: '')
 option('xdg-foreign', type: 'feature', value: 'auto', description: 'Enable xdg-foreign protocol')
+option('backends', type: 'combo', choices: ['auto', 'drm', 'noop', 'headless', 'libinput', 'x11', 'wayland'],
+       description: 'Select the backend support')
```
Comment (emersion) [inline meson_options.txt:11]: This is missing an option to toggle the GLES2 renderer.

**Review by emersion — CHANGES_REQUESTED** [review]
Thanks for working on this, here's a first round of comments.

I really like the new `backends` option, that's a good idea.

I don't think it makes sense to have an option to build without GBM, since the renderers/backends will just fail in this case. We might want to wait for other allocators to kick in before doing that.

In general it doesn't seem like the commits are properly split up. The first commit uses `#ifdef`s with symbols introduced later on, for instance.

File: include/wlr/config.h.in (hunk 1 above)
Comment (hizukiayaka) [reply include/wlr/config.h.in:22]: `WLR_HAS_LIBSEAT`
There is a library that inside backend session, the only reason I made BACKEND suffix is that those macro are only used
for backend. Well if you want that, I can agree with you, but it just make that macro look too so long.

File: meson_options.txt (hunk 3 above)
Comment (hizukiayaka) [reply meson_options.txt:11]: Well, it is not harm to keep such stub code, anyway it would link to a GLES2 library anyway.
If that platform doesn't have a GPU, it won't be able to find a library, the link check would disable the renderer there.

Comment (emersion) [reply meson_options.txt:11]: Let me rephrase this: we can't have `dependency('glesv2', optional: true)` without a way for the user to explicitly enable or disable it. Some users may have the library installed but may still want to disable the GLES2 renderer. Auto-detection is nice for end users, but cause issues in CI and automated build environments.

File: include/wlr/config.h.in (hunk 1 above)
Comment (emersion) [reply include/wlr/config.h.in:22]: I'd prefer to have the `_BACKEND` prefix for consistency.

File: backend/x11/meson.build [hunk 4]
```diff
@@ -10,16 +10,16 @@ x11_required = [
 ]
 
 msg = []
-if get_option('x11-backend').enabled()
-	msg += 'Install "@0@" or pass "-Dx11-backend=disabled" to disable it.'
-endif
-if not get_option('x11-backend').disabled()
-	msg += 'Required for X11 backend support.'
+
+if 'x11'.contains(backends)
```
Comment (kennylevinsen) [inline backend/x11/meson.build:14]: This looks like it will evaluate to true if `backends` is equal to or a substring of "x11". I imagine you wanted it to be true if one of the elements in `backends` was equal to "x11". Am I missing some magic here?

File: backend/meson.build [hunk 5]
```diff
@@ -1,11 +1,13 @@
 wlr_files += files('backend.c')
 
-subdir('drm')
-subdir('headless')
-subdir('libinput')
+backends = get_option('backends')
+
+foreach backend : ['drm', 'noop', 'headless', 'libinput', 'x11', 'wayland']
+  if [backend, 'auto'].contains(backends)
```
Comment (kennylevinsen) [inline backend/meson.build:6]: This looks like it will evaluate to true if `backends` in its entirety is equal to `backend` or to "auto". I imagine you wanted it to be true if one of the *elements* in `backends` was equal to `backend` or "auto". Am I missing some magic here?

File: backend/headless/backend.c [hunk 6]
```diff
@@ -220,21 +225,24 @@ struct wlr_backend *wlr_headless_backend_create(struct wl_display *display) {
...
 
+#if WLR_HAS_GBM_SUPPORT
 	struct wlr_gbm_allocator *gbm_alloc = wlr_gbm_allocator_create(drm_fd);
 	if (gbm_alloc == NULL) {
 		wlr_log(WLR_ERROR, "Failed to create GBM allocator");
 		close(drm_fd);
 		goto error_dup;
 	}
+	base_allocator = &gbm_alloc->base;
+#endif
```
Comment (kennylevinsen) [inline backend/headless/backend.c:236]: I don't normally mind `#ifdef`'s, but I feel like there's a lot of repetition throughout this change.

Options:
- A private, shared helper (`wlr_gbm_allocator_try_create`?) where the `#ifdefs` go.
- Something like a `wlr_allocator_autocreate`
- Let `wlr_gbm_allocator_create` exist but fail when GBM is unavailable - this approach could also work for other components, such as libinput and gles.

@emersion what do you think?

**Review by kennylevinsen — CHANGES_REQUESTED** [review]
Thanks for the PR! Here's a few comments.

It should also be noted that emersion's last comment remains unresolved.

File: backend/headless/backend.c (hunk 6 above)
Comment (emersion) [reply backend/headless/backend.c:236]: > Let wlr_gbm_allocator_create exist but fail when GBM is unavailable - this approach could also work for other components, such as libinput and gles.

This doesn't let us make a difference between "dependency is missing" and "something went wrong".

> Something like a wlr_allocator_autocreate

This seems like a good idea. There are interactions with https://github.com/swaywm/wlroots/pull/2737#discussion_r602707670 here: the `wlr_allocator_autocreate` caller doesn't know anymore whether the allocator will use shm or DMA-BUFs. The caller also can't guarantee the allocator and the renderer will work fine together. If we keep the new function private for now, we could sort out these issues later.

Comment (kennylevinsen) [reply backend/headless/backend.c:236]: > This doesn't let us make a difference between "dependency is missing" and "something went wrong".

That's true. I imagine the states would only be unavailable, broken dependency and allocation error. If proceeding to the next available allocator is acceptable, then we could just log the cause from the _create method.

> [regarding wlr_allocator_autocreate] If we keep the new function private for now, we could sort out these issues later.

I'd ACK that approach. @hizukiayaka, does that work for you?

Comment (hizukiayaka) [reply backend/headless/backend.c:236]: `wlr_allocator_autocreate` would be used as we all agreed, neither gbm allocator nor drm dumb header files would be exported anymore.
Once we don't the part how to deal with the buffer its allocate or even the part the buffer it imported, we could remove all the `#ifdef` here

File: render/wlr_renderer.c [hunk 7]
```diff
@@ -1,9 +1,17 @@
 #include <assert.h>
 #include <stdbool.h>
 #include <stdlib.h>
+#include <wlr/config.h>
+
+#if WLR_HAS_GBM_SUPPORT
```
Comment (emersion) [inline render/wlr_renderer.c:6]: `WLR_HAS_GBM_SUPPORT` doesn't exist in this commit. It's introduced in a later commit.

Comment (emersion) [reply render/wlr_renderer.c:6]: Ah, no, it's added in `features += { 'gbm-support': true }`.

I'd prefer to defer the GBM issue to a later commit.

File: meson_options.txt [hunk 8]
```diff
@@ -7,3 +7,5 @@ option('x11-backend', type: 'feature', value: 'auto', description: 'Enable X11 b
 option('examples', type: 'boolean', value: true, description: 'Build example applications')
 option('icon_directory', description: 'Location used to look for cursors (default: ${datadir}/icons)', type: 'string', value: '')
 option('xdg-foreign', type: 'feature', value: 'auto', description: 'Enable xdg-foreign protocol')
+option('renderers', type: 'combo', choices: ['auto', 'gles2'], description:
```
Comment (emersion) [inline meson_options.txt:9]: I think this should be `array` instead of `combo`, so that we can add more renderers later.

File: render/gles2/meson.build [hunk 9]
```diff
@@ -1,3 +1,12 @@
+glesv2 = dependency('glesv2', required : false)
```
Comment (emersion) [inline render/gles2/meson.build:1]: If the user has explicitely selected the GLES2 backend via the `renderers` option, then the `glesv2` dependency is required.

File: render/meson.build [hunk 10]
```diff
@@ -1,13 +1,33 @@
+renderers = get_option('renderers')
+egl = dependency('egl', required : false)
```
Comment (emersion) [inline render/meson.build:2]: EGL is required for the GLES2 renderer.

File: render/meson.build [hunk 11]
```diff
@@ -1,13 +1,33 @@
...
+  wlr_files += files(
+	  'egl.c',
+	  'gbm_allocator.c',
+  )
+  features += { 'gbm-support': true }
+  wlr_deps += [egl, gbm]
+
+# Actually, a GLES renderer could still render into a memory buffer
+# while those headless device would work in this way
+  if ['gles2', 'auto'].contains(renderers)
```
Comment (emersion) [inline render/meson.build:28]: The GLES2 renderer doesn't depend on GBM, I think.

File: render/meson.build [hunk 12]
```diff
@@ -1,13 +1,33 @@
+renderers = get_option('renderers')
+egl = dependency('egl', required : false)
+gbm = dependency('gbm', version: '>=17.1.0', required : false)
```
Comment (emersion) [inline render/meson.build:3]: Like I said below, I'd prefer to always require GBM for now, and make it optional later, once we sort out the issues with the DRM backend.

File: backend/backend.c [hunk 13]
```diff
@@ -198,6 +207,7 @@ static struct wlr_backend *attempt_backend_by_name(struct wl_display *display,
 		return attempt_headless_backend(display);
 	} else if (strcmp(name, "noop") == 0) {
 		return attempt_noop_backend(display);
+#if WLR_HAS_DRM_BACKEND
```
Comment (emersion) [inline backend/backend.c:210]: I think we can move this `#if` to only enclose the `attempt_drm_backend` call.

File: backend/drm/meson.build [hunk 14]
```diff
@@ -1,3 +1,7 @@
+if gbm.found() == false
```
Comment (emersion) [inline backend/drm/meson.build:1]: `not gbm.found()`

… But let's keep the GBM stuff for later

File: meson_options.txt [hunk 15]
```diff
@@ -3,9 +3,10 @@ option('logind-provider', type: 'combo', choices: ['auto', 'systemd', 'elogind']
 option('libseat', type: 'feature', value: 'auto', description: 'Enable support for rootless session via libseat')
 option('xcb-errors', type: 'feature', value: 'auto', description: 'Use xcb-errors util library')
 option('xwayland', type: 'feature', value: 'auto', yield: true, description: 'Enable support for X11 applications')
-option('x11-backend', type: 'feature', value: 'auto', description: 'Enable X11 backend')
 option('examples', type: 'boolean', value: true, description: 'Build example applications')
 option('icon_directory', description: 'Location used to look for cursors (default: ${datadir}/icons)', type: 'string', value: '')
 option('xdg-foreign', type: 'feature', value: 'auto', description: 'Enable xdg-foreign protocol')
 option('renderers', type: 'combo', choices: ['auto', 'gles2'], description:
        'Select the built-in renderer')
+option('backends', type: 'combo', choices: ['auto', 'drm', 'noop', 'headless', 'libinput', 'x11', 'wayland'],
```
Comment (emersion) [inline meson_options.txt:11]: This should be an `array`, or else this only allows to select a single backend.

File: meson_options.txt (hunk 8 above)
Comment (hizukiayaka) [reply meson_options.txt:9]: Got it, yes it should be

File: render/gles2/meson.build (hunk 9 above)
Comment (hizukiayaka) [reply render/gles2/meson.build:1]: but if only the auto is selected, we can't make it be required here. I could make it throw an error if the gles2 is selected

File: render/meson.build (hunk 10 above)
Comment (hizukiayaka) [reply render/meson.build:2]: For the current code, it is, in theory  no. There is opengl offscreen, that would be used for those headless backend

File: backend/drm/meson.build (hunk 14 above)
Comment (hizukiayaka) [reply backend/drm/meson.build:1]: I fix it in the other place but how could it pas my test, I did do compiling test

File: meson_options.txt (hunk 15 above)
Comment (hizukiayaka) [reply meson_options.txt:11]: yes, my fault

File: render/gles2/meson.build (hunk 9 above)
Comment (emersion) [reply render/gles2/meson.build:1]: This should do it:

```
glesv2 = dependency('glesv2', required: get_option('renderers').contains('gles2'))
```

aka. required only if the GLES2 renderer was explicitly enabled by the user.

File: render/meson.build (hunk 12 above)
Comment (hizukiayaka) [reply render/meson.build:3]: That is right, I thought we use the surface from gbm, that is what weston does. But wlroots didn't use gbm surface or EGL surface, it would also using EGLImage -> Renderbuffer -> FBO
I would move egl and gles2 into the same directory and EGL doesn't depend on gbm.

**Comment by emersion** [comment]
I'm not sure moving `egl.c` is a good idea. It's true that right now it's only used in the GLES2 renderer, but it may be used in other GL renderers in the future. But I guess my main concern with this change is that the include paths and the C source paths are not consistent anymore: `include/render/egl.h` and `include/wlr/render/egl.h` are not consistent with `render/gles2/egl.c`.

File: render/meson.build [hunk 16]
```diff
@@ -1,13 +1,27 @@
...
 	'drm_format_set.c',
-	'gbm_allocator.c',
 	'pixel_format.c',
 	'swapchain.c',
 	'wlr_renderer.c',
 	'wlr_texture.c',
 )
+wlr_deps += drm_headers
+
+if gbm.found()
```
Comment (emersion) [inline render/meson.build:17]: I'd prefer to split that GBM change into a separate commit, as said elsewhere.

I can do the commit splitting if that sounds complicated to you.

File: backend/x11/backend.c [hunk 17]
```diff
@@ -569,13 +569,13 @@ struct wlr_backend *wlr_x11_backend_create(struct wl_display *display,
 		goto error_event;
 	}
 
-	struct wlr_gbm_allocator *gbm_alloc = wlr_gbm_allocator_create(drm_fd);
-	if (gbm_alloc == NULL) {
-		wlr_log(WLR_ERROR, "Failed to create GBM allocator");
+	alloc = wlr_allocator_create_with_drm_fd(drm_fd);
```
Comment (emersion) [inline backend/x11/backend.c:572]: No need for this intermediate variable, can directly assign to `x11->allocator`

File: backend/wayland/backend.c [hunk 18]
```diff
@@ -439,15 +436,13 @@ struct wlr_backend *wlr_wl_backend_create(struct wl_display *display,
 		goto error_drm_fd;
 	}
 
-#if WLR_HAS_GBM_SUPPORT
-	struct wlr_gbm_allocator *gbm_alloc = wlr_gbm_allocator_create(drm_fd);
-	if (gbm_alloc == NULL) {
-		wlr_log(WLR_ERROR, "Failed to create GBM allocator");
+	alloc = wlr_allocator_create_with_drm_fd(drm_fd);
+	if (alloc == NULL) {
```
Comment (emersion) [inline backend/wayland/backend.c:440]: No need for this intermediate variable, can directly assign to `wl->allocator`

File: backend/headless/backend.c [hunk 19]
```diff
@@ -254,7 +248,7 @@ struct wlr_backend *wlr_headless_backend_create_with_renderer(
 		struct wl_display *display, struct wlr_renderer *renderer) {
 	wlr_log(WLR_INFO, "Creating headless backend with parent renderer");
 
-	struct wlr_allocator *base_allocator = NULL;
+	struct wlr_allocator *alloc = NULL;
```
Comment (emersion) [inline backend/headless/backend.c:251]: Can declare this variable below when it's first used.

File: backend/headless/backend.c [hunk 20]
```diff
@@ -205,7 +202,7 @@ static int open_drm_render_node(void) {
 struct wlr_backend *wlr_headless_backend_create(struct wl_display *display) {
 	wlr_log(WLR_INFO, "Creating headless backend");
 
-	struct wlr_allocator *base_allocator = NULL;
+	struct wlr_allocator *alloc = NULL;
```
Comment (emersion) [inline backend/headless/backend.c:205]: Can declare this variable below when it's first used.

**Comment by emersion** [comment]
This PR needs to be rebased on top of master because of conflicts.

**Comment by emersion** [comment]
I've rejiggered and cleaned up this PR in this branch: https://github.com/emersion/wlroots/commits/optional-backends-renderers

The "render: support platforms without GBM" commit is missing, because the later patches invalidate it and I don't want to merge it first.

**Comment by emersion** [comment]
I've rebased and merged the GLES2 patch: https://github.com/swaywm/wlroots/commit/ed1924800de905f5462c8cc6e36fbc60b69d3030. The Pixman renderer hasn't been made optional, because we need the Pixman dependency anyways.

The patch about backends needs more work. I've rebased it and cleaned it up a bit in my branch: https://github.com/emersion/wlroots/commits/optional-backends-renderers

I don't think disabling headless, noop and wayland backends is desirable, because that doesn't reduce the number of dependencies. To address the GBM dependency issue, a plan is laid out in https://github.com/swaywm/wlroots/issues/2861. Now that the Pixman renderer has been merged, we could allow building wlroots without the GBM allocator and GLES2.

**Comment by hizukiayaka** [comment]
@emersion oh, I should read your post first
disable some backends may just want to save more space for those embedded device, headless could lead to a complex implementation likes weston does
I am worrying nested wayland compositor may be enabled by wrong environment value or maybe it uses a different protocol scanner

**Comment by hizukiayaka** [comment]
As I explained, if let auto_features bother those option which is an array, it would become too  complex, the value 'auto' is just a quick for them, if they want to disable something, they should set an empty arrary, auto_features won't help them there.

**Comment by emersion** [comment]
Disagree. We should at least error out if `auto_features=enabled`/`disabled` and the array has "auto" in it. Or better, do the right thing.

**Comment by hizukiayaka** [comment]
> Disagree. We should at least error out if `auto_features=enabled`/`disabled` and the array has "auto" in it. Or better, do the right thing.
I would agree to throw an error in the case your described, I don't want to invoke the other part for the 'auto' quirk here.

**Comment by emersion** [comment]
https://github.com/swaywm/wlroots/pull/3058 has been merged. This PR needs a rebase.

**Comment by emersion** [comment]
Most of the things this PR does have been merged.
